lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManager.java

```diff
@@ -61,29 +106,61 @@
    *
    * @param ids the evaluator IDs of the evaluators that are expected to have survived driver restart.
    */
-  void setPreviousEvaluatorIds(final Set<String> ids);
+  public void setPreviousEvaluatorIds(final Set<String> ids) {
+    if (this.state != DriverRestartState.RestartInProgress) {
+      previousEvaluators.addAll(ids);
+      this.state = DriverRestartState.RestartInProgress;
+    } else {
+      final String errMsg = "Should not be setting the set of expected alive evaluators more than once.";
+      LOG.log(Level.SEVERE, errMsg);
+      throw new DriverFatalRuntimeException(errMsg);
+    }
+  }
 
   /**
    * @return the IDs of the Evaluators from a previous Driver that have checked in with the Driver
    * in a restart situation.
    */
-  Set<String> getRecoveredEvaluatorIds();
+  public Set<String> getRecoveredEvaluatorIds() {
+    return Collections.unmodifiableSet(this.previousEvaluators);
+  }
 
   /**
    * Indicate that this Driver has re-established the connection with one more Evaluator of a previous run.
    * @return true if the driver restart is completed.
    */
-  boolean evaluatorRecovered(final String id);
+  public boolean evaluatorRecovered(final String evaluatorId) {
```

Comment:
Using onRecoverEvaluatorIsRestartComplete, because it returns true if restart completes instead of returning whether or not this is a recoveredEvaluator. This is to avoid a race condition in EvaluatorManager between onEvaluatorRecovered and isRestartCompleted, which may potentially cause the restart complete handler to be invoked more than once.